### DISCUSSION TITLE
Assignment using Grade from Log not shelving all variables

### DISCUSSION THREAD

**Discussion by lebrown** [ucbds-infra/otter-grader#871] [Q&A] [answered]
I am working on an assignment notebook using the "Grade from Log" functionality. 

I have the configuration set with: 
```
save_environment: true 
generate: 
    grade_from_log: true
```
I noticed in my third question that 3 variables are not be shelved out to the log file.  Therefore, when I use my solutions as a submission in Gradescope I have test cases failing with a Name Error.  

For example: 
```
        Trying:
            isinstance(product_info_out, dict)
        Expecting:
            True
        **********************************************************************
        Line 1, in q3 1
        Failed example:
            isinstance(product_info_out, dict)
        Exception raised:
            Traceback (most recent call last):
              File "/root/miniforge3/envs/otter-env/lib/python3.10/doctest.py", line 1350, in __run
                exec(compile(example.source, filename, "single",
              File "<doctest q3 1[0]>", line 1, in <module>
                isinstance(product_info_out, dict)
            NameError: name 'product_info_out' is not defined
```

I can see in the log file that the variable `product_info_out` is not being stored. 

```
from otter.check.logs import Log
log = Log.from_file(".OTTER_LOG")
entry = log.get_question_entry("q3")
env = entry.unshelve()
env.keys()
```
```output
dict_keys(['__name__', '__doc__', '__package__', '__loader__', '__spec__', '_ih', '_oh', '_dh', 'In', 'Out', 'open', '__', '___', '__session__', '_i', '_ii', '_iii', '_i1', 'BeautifulSoup', 'StringIO', 'GS', '_i2', 'site', 'q1', '_i3', 'q1datasets', 'q1_num_datasets', '_i4', '_4', '_i5', 'q2', '_i6', 'names', '_6', '_i7', '_7', '_i8', 'extract_book_links', '_i9', 'get_product_info', '_i10', 'scrape_books', '_i11', 'extract_book_info_out', 'extract_book_url', 'scrape_books_out', '_i12', 'scrape_books_out2', '_i13']
```

Any thoughts on why this is happening?

**Comment by chrispyles** [comment]
Nothing jumps out at me here; can you share the full notebook? Feel free to PM it to me on Slack (link in README) if you don’t want it shared publicly.

**Reply by lebrown** [reply]
Thanks! I just sent you a message on Slack!

**Reply by chrispyles** [reply]
It looks like the `product_info_out` variable throws an error when dill tries to serialize it:

```
RecursionError: maximum recursion depth exceeded while calling a Python object
```

(Note: you can see the variable names in the environment that weren't shelved in `<log entry>.not_shelved`)

It's getting thrown while trying to serialize the `Description` value. Unfortunately I can't think of a good solution to this other than removing it from `product_info_out`.

**Reply by lebrown** [reply, marked as answer]
Thanks for looking into to this.  I noticed in the `Description` value, I had not actually used the `.text` on the HTML parsing.  
After adding this in, there doesn't seem to be an issue with dill serializing the object. 

Again thanks for you help!  This pointed me in the right direction to get it fixed.
